### PR TITLE
SONARSLANG-5 Rule: Redundant Boolean Literals

### PR DESCRIPTION
**Pull request by andrea-guarino-sonarsource**

### CODE REVIEW

File: slang-checks/src/test/resources/com/sonarsource/slang/checks/BooleanLiteral.slang [hunk 1]
```diff
@@ -0,0 +1,26 @@
+x == true;                                   // Noncompliant {{Remove the unnecessary Boolean literal.}}
+//   ^^^^
+x == false;                                  // Noncompliant {{Remove the unnecessary Boolean literal.}}
+//   ^^^^^
+x != true;                                   // Noncompliant {{Remove the unnecessary Boolean literal.}}
+x != false;                                  // Noncompliant {{Remove the unnecessary Boolean literal.}}
+true == x;                                   // Noncompliant {{Remove the unnecessary Boolean literal.}}
+false == x;                                  // Noncompliant {{Remove the unnecessary Boolean literal.}}
```
Comment (pynicolas) [inline slang-checks/src/test/resources/com/sonarsource/slang/checks/BooleanLiteral.slang]: I don't see any case where `==`  or `!=` is used without a boolean literal.

File: slang-checks/src/test/resources/com/sonarsource/slang/checks/BooleanLiteral.slang [hunk 2]
```diff
@@ -0,0 +1,26 @@
...
+!false;                                      // Noncompliant {{Remove the unnecessary Boolean literal.}}
+false && foo;                                // Noncompliant {{Remove the unnecessary Boolean literal.}}
+x || true;                                   // Noncompliant {{Remove the unnecessary Boolean literal.}}
+
+!x                                           // OK
+x || foo;                                    // OK
+
+x = if (foo) y else false;                   // Noncompliant {{Remove the unnecessary Boolean literal.}}
+x = if (foo) y else true;                    // Noncompliant {{Remove the unnecessary Boolean literal.}}
+x = if (foo) y else true;                    // Noncompliant {{Remove the unnecessary Boolean literal.}}
```
Comment (pynicolas) [inline slang-checks/src/test/resources/com/sonarsource/slang/checks/BooleanLiteral.slang]: Is there a difference with the previous line?

File: slang-checks/src/test/resources/com/sonarsource/slang/checks/BooleanLiteral.slang [hunk 3]
```diff
@@ -0,0 +1,26 @@
...
+x = if (foo) true else y;                    // Noncompliant {{Remove the unnecessary Boolean literal.}}
+x = if (foo) false else y;                   // Noncompliant {{Remove the unnecessary Boolean literal.}}
+x = if (foo) false                           // Noncompliant {{Remove the unnecessary Boolean literal.}}
```
Comment (pynicolas) [inline slang-checks/src/test/resources/com/sonarsource/slang/checks/BooleanLiteral.slang]: I don't know what would the semantic of this statement in slang, but it's not valid in kotlin.

File: sonar-slang-plugin/src/main/resources/org/sonar/l10n/kotlin/rules/kotlin/S1125.html [hunk 4]
```diff
@@ -0,0 +1 @@
+<p>Redundant Boolean literals should be removed from expressions to improve readability.</p>
```
Comment (pynicolas) [inline sonar-slang-plugin/src/main/resources/org/sonar/l10n/kotlin/rules/kotlin/S1125.html]: For this rule, I believe we should try to provide some code examples. In some cases, what this rule catches is not obvious.

**Comment by andrea-guarino-sonarsource** [comment]
Issue https://jira.sonarsource.com/browse/SONARSLANG-28 has been create to handle ternary operator

File: slang-checks/src/main/java/com/sonarsource/slang/checks/BooleanLiteralCheck.java [hunk 5]
```diff
@@ -0,0 +1,71 @@
...
+import com.sonarsource.slang.checks.api.InitContext;
+import com.sonarsource.slang.checks.api.SlangCheck;
+import java.util.Arrays;
+import java.util.List;
+import java.util.Optional;
+import org.sonar.check.Rule;
+
+@Rule(key = "S1125")
+public class BooleanLiteralCheck implements SlangCheck {
+  private static final List<String> BOOLEAN_LITERALS = Arrays.asList("true", "false");
```
Comment (pynicolas) [inline slang-checks/src/main/java/com/sonarsource/slang/checks/BooleanLiteralCheck.java]: I'm not sure that these values will be valid for all languages. Strictly speaking, we should define BooleanLiteralTree. Let's keep it like that for the moment...

File: slang-checks/src/main/java/com/sonarsource/slang/checks/BooleanLiteralCheck.java [hunk 6]
```diff
@@ -0,0 +1,71 @@
...
+        getBooleanLiteral(unaryExprTree.operand())
+          .ifPresent(booleanLiteral -> ctx.reportIssue(booleanLiteral, MESSAGE));
+      }
+    });
+  }
+
+  private static Optional<LiteralTree> getBooleanLiteral(Tree... trees) {
+    return Arrays.stream(trees)
+      .filter(tree -> tree instanceof LiteralTree && BOOLEAN_LITERALS.contains(((LiteralTree) tree).value()))
+      .map(LiteralTree.class::cast)
```
Comment (pynicolas) [inline slang-checks/src/main/java/com/sonarsource/slang/checks/BooleanLiteralCheck.java]: Do we need that line?

File: slang-checks/src/test/resources/com/sonarsource/slang/checks/BooleanLiteral.slang (hunk 2 above)
Comment (andrea-guarino-sonarsource) [reply slang-checks/src/test/resources/com/sonarsource/slang/checks/BooleanLiteral.slang]: I think I duplicate this line by mistake

**Pull request merged**